[먹이사슬] 미션 구현 (신윤철)

Shin-Yun-Cheol opened the pull request:
## 💡 Summary
구현한 기능
- 먹이사슬을 그래프로 저장
- stream 사용해보기

## ✅ CheckList

- [x] - 기본 기능 구현
- [x] - 응용 기능 구현
- [ ] - 심화 기능 구현
- [ ] - 테스트 케이스 만들어보기
- [x] - stream 써보기
- [x] - mvc 패턴 이용하기

## 신경써서 구현한 점 / 어려웠던 점
- mvc 패턴 공부해서 적용해보기
- stream 이용해보기
- 그래프 이용해보기 

## 📚 Retrospective (YWT)

### What I did
-  MVC를 통해 의존성 분리
- stream이용해서 좀 더 편리하게 반복문 이용해보기

### What I Learned
-  MVC 패턴과 의존성 학습
- 그래프 자료구조를 프로그램 짜는 데 이용해보는 법 학습
- stream 이용법 학습
- 

### What I will do next
- 알고리즘 더 공부해보기
- 테스트 케이스도 만들어보기


## ✨ Impression
이번 미션은  mvc 패턴을 처음 적용해보려고 하다 보니 뭔가 기본부터 쉽지 않았던 느낌이 든다. 그래도 지난 주 미션보다 mvc 패턴을 이용해서 코드가 훨씬 깔끔해 보이게 만들 수 있던 거 같다. 
문제를 잘 이해하지 못한 이유도 있겠지만 bfs, dfs 알고리즘을 프로그래밍 하는 것에 대한 지식이 부족해서 어떻게 구현해야할 지 막막했고, 그래프도 잘 안써봐서 더 어렵게 느껴졌지만 mvc 나 stream 을 직접 이용해보고 그래프도 조금이나마 이용해볼 수 있어서 많이 배울 수 있었다. 
다만 심화까지 구현하지 못해서 아쉬웠고 테스트 케이스나 예외처리 하는 것도 다음번엔 더 해봐야겠다는 생각이 들었다.

Shin-Yun-Cheol committed:
feat : 기본 단계 물고기 정보 가져오기까지 구현

MVC패턴을 적용해서 각각 class만들고 입력 받는 걸 파싱해서 정보 가져오는 것 까지 구현했습니다.

Shin-Yun-Cheol committed:
feat : 기본 단계 구현완료

util에 SurvivalSimulator.java을 추가해서 시뮬레이션 하는 클래스를 만들고 Controller에 연결하고 Application.java에 객체 생성해서 실행할 수 있도록 했습니다. FishInputParser.java에 예외처리 하는 코드 추가했습니다.

Shin-Yun-Cheol committed:
feat : 응용 단계 구현완료

FishController.java에 predationGraph를 만들어서 먹이사슬을 그래프로 표현했습니다. List를 하나 만들고 stream을 이용해 정렬해서 보관했습니다.
SurvivalSimulator.java에서는 tropthicLevel이 아니라 먹이사슬 id로 연산할 수 있게 로직을 변경했습니다.

seulnan commented on src/main/java/mission/util/SurvivalSimulator.java:
이 파일은 재사용성보다는 도메인 규칙과 로직 흐름이 강하게 반영되어 있기 때문에 유틸보다는 서비스 레이어에 위치하는 게 적절해 보여요.

1. 컨트롤러는 사용자 요청을 받아 서비스를 호출하고, 
2. 서비스는 비즈니스 로직을 처리한 뒤 결과를 리턴하면,
3. 그 결과가 포맷팅이나 변환이 필요한 경우, 그때 유틸을 통해 파싱하거나 매핑해서 
4. 이 결과를 뷰에 전달하는 구조

가 이상적이라고 생각해요.

추가적으로 controller와 서비스 차이점, 그리고 서비스와 유틸의 차이점 위주로 공부해보시면 좋을 것 같아요!

seulnan replied:
어쩌라고

seulnan replied:
[예시링크](https://brotherdan.tistory.com/42)

Shin-Yun-Cheol replied:
정리하자면 이렇게 알아봣습니다!

Controller : 사용자 요청을 받아서 이를 적절한 로직(Service)로 전달하는 통로 같은 존재
Service : 비즈니스 로직같은 구체적인 작업을 수행
Util : 로직을 수행하는데 필요한 데이터 가공과 조작

MVC패턴을 적용하라고 해서 진짜 M, V, C 만 있는 줄 알고 미션을 수행했는데 그게 아니라 세부적으로 많은 것들이 있었네요 .
Controller와 Service에 대해 찾다보니 외부와 연결하는 Repository라는 내용도 많이 나오는데 아직 공부할 게 많은 것 같습니다..

seulnan replied:
좋습니다. Repository는 스프링에서 DB에 연결할 때 사용하는 것들이라 주로 API 설계할 때 사용합니다. 
결론: 우리 자바스터디에서는 Repository 역할을 하는 정도의 코드를 만들순 있겠지만 실제로 DB를 연결할 일은 없습니다.

seulnan commented on src/main/java/mission/controller/FishController.java:
```diff
@@ -0,0 +1,60 @@
+        SurvivalSimulator simulator = new SurvivalSimulator(predationGraph);
+        int days = simulator.simulate(fishWithCounts);
+
+        System.out.println(days + "일간 생존했습니다.");
```
현재 출력 관련 코드가 Controller 내부에 섞여 있는데, view는 실제 프로젝트 진행 중 가장 자주 변경되는 영역 중 하나입니다. (우리스터디는 좀 예외인 것 같긴하지만)

만약 이 View 로직이 여러 레이어에 흩어져 있으면, 출력 포맷을 변경할 때 의존된 모든 코드가 함께 영향을 받아 시스템 전체가 와장창 깨질 수 있어요.

따라서 View는 Controller에서만 직접 의존하고, 하위 도메인이나 서비스에서는 직접 의존하지 않도록 계층을 철저히 분리하는 것이 좋습니다.

참고로 Spring 환경에서는 View뿐만 아니라 Request/Response DTO도 변경이 잦은 영역이라 도메인 모델과는 철저히 분리하고, 컨트롤러를 중심으로만 사용되도록 의존성을 제한하는 것이 일반적인 설계 방향입니다.

참고 글
https://hyunsb.tistory.com/92

Shin-Yun-Cheol replied:
sout으로 처리하던 것들을 OutputView 클래스를 하나 만들어서 그곳에서 처리할 수 있게 만들었습니다.
컨트롤러는 모델과 서비스를 잇는 역할이고 그 외의 다른 역할들은 서비스와 클래스들에 맡기는 것으로 이해했습니다. 
공부할수록 더 세분화 하고 한 클래스에서 정의하는게 적고 여러 클래스를 잘 잇는게 중요하다는 생각이 드네요

seulnan commented on src/main/java/mission/controller/FishController.java:
컨트롤러가 너무 많은 책임을 가지고 있는것 같아요.
(인풋파서의 리턴값을 매핑해서 리스트로 변환하는 fishWithCounts라든가, 그래프 정의, 시뮬레이션까지)
도메인 별 service로 분리하거나 메소드를 분리하는 식으로 코드변경이 필요해보입니다.

Q. 왜 컨트롤러가 많은 책임을 지면 안될까요?

Shin-Yun-Cheol replied:
[찾아보니](https://inpa.tistory.com/entry/OOP-%F0%9F%92%A0-%EC%95%84%EC%A3%BC-%EC%89%BD%EA%B2%8C-%EC%9D%B4%ED%95%B4%ED%95%98%EB%8A%94-SRP-%EB%8B%A8%EC%9D%BC-%EC%B1%85%EC%9E%84-%EC%9B%90%EC%B9%99) SRP(Single Responsibility Principle)(단일 책임 원칙)이라는게 있는데..  
기능 변경이 일어났을 때의 파급 효과를 위해 클래스 하나에 기능 하나를 넣는게  좋고, 변경하는 이유가 하나인게 좋다고 합니다.

제 코드는 fishWithCounts매핑, 그래프 생성, 시뮬레이션 실행, 결과 출력까지 있어서 완전 위반하고 있었는데.. 

그래프는 클래스로 , 매핑은 메서드로, 출력은 OutputView로 리팩토링 해보겠습니다!

[커밋링크](https://github.com/allrounder-backend/mission-food-chain/pull/3/commits/23593b5fa555a65372428ca2ce0697e3975326e3)

seulnan commented on src/main/java/mission/Application.java:
```diff
@@ -1,9 +1,34 @@
         //TODO: 미션 구현
+
+        List<Fish> data = List.of(
+                new Fish(1, "플랑크톤", 1),
```
팩토리 패턴을 한번 공부해보고 적용해보시면 좋을듯해요!

Shin-Yun-Cheol replied:
팩토리 패턴에는 Simple Factory, Factory Method, Abstract Factory 세 종류가 있는 것으로 찾아봤습니다. 

- Simple Factory에서는 객체를 생성하는 클래스를 따로 만들어서 if-else, swich문 등으로 모든 객체를 만드는 방법입니다. 단순히 객체 생성을 분리하는 방법이라고 이해했습니다. 단점으로는 if, switch 남발이 있습니다.

- Factory Method는 interface를 만들고 서브클래스에서 어떤 객체를 생성할지 결정할 수 있게 하는 방법입니다. 그렇기 때문에 나중에 만들고 싶은 객체가 늘어나도 서브클래스만 더 만들면 되기 때문에 확장이 유용합니다. 이는 OCP원칙(Open Close Principle)(확장과 변경)을 따르게 합니다.

- Abstract Factory는 여러 객체를 세트로 일관되게 만들어야 할 때 이용합니다. 예를 들어 제품군 전체를 만드는데 PhoneFactory에서 Phone과 OS를 같이 만든다거나, UI를 만들 때 버튼과 체크박스를 같이 만드는 상황을 들 수 있습니다. 단점으로는 새로운 종류를 제공하기 어렵다는 점입니다. Factory에 추상 메소드가 하나 더 추가되면 모든 서브 구현체를 다시 수정해야합니다.


코멘트 주신 부분은 뒤에 두 방법보단 Simple Factory 정도를 적용해볼 수 있을 것 같다고 생각하여 그렇게 수정할 예정입니다. 이는 코드의 생성 책임을 분리할 수 있을 것입니다.

[이 글](https://cjw-awdsd.tistory.com/54)을 참고하여 작성했습니다.

seulnan replied:
네네 저도 심플팩토리를 적용하라는 의미에서 단 리뷰였어요! 좋습니다. if switch를 남발하지않기 위해 여러가지 해결방법이 있으니 이것도 알아보시면 좋을듯해요

seulnan commented on src/main/java/mission/controller/FishController.java:
```diff
@@ -0,0 +1,60 @@
+                                .collect(Collectors.toList());
+
+        Map<Integer, List<Integer>> predationGraph = new HashMap<>();
+        predationGraph.put(2, List.of(1));
```
이렇게 하드코딩하는 것보다 enum을 한번 활용해보시는것도 좋아보여요.
enum은 왜 쓰는지, 어떻게 쓰는건지를 공부해보셔요.

Shin-Yun-Cheol replied:
오... enum을 클래스처럼 쓸 수도 있군요 !  
Exception관리 할 때도 썼지만 이런 변하지 않는 상태나 데이터 관리할 때 좋은 거 같아요.
추가할 것 있으면 enum에 추가만 하면 되고요!!

seulnan commented on src/main/java/mission/controller/FishController.java:
```diff
@@ -0,0 +1,60 @@
+                            Fish fish = fishData.stream()
+                                    .filter(f-> f.getName().equals(entry.getKey()))
+                                    .findFirst()
+                                    .orElseThrow(()-> new IllegalArgumentException("존재하지 않음"));
```
예외처리 관련한것은 validate, 에러메시지는 enum을 활용하고, exception객체를 이용하는 것은 어떨까요?? 

검증로직과 에러객체들은 추후에 재사용가능성이 높고, 무엇을 검사하고 어떤 규칙이 위배되었는지 파악함으로써 에러흐름이 명확하게 드러나서 유지보수성, 도메인 구분, 예외 핸들링면에서 좋습니다.

Shin-Yun-Cheol replied:
exception패키지를 만들고 enum을 통해 에러메시지 정의했습니다. InvalidException이라는 클래스를 만들어서 예외 객체 생성할 수 있게 했습니다. 또한 OutputView에 에러메시지 출력하는 메소드 만들어서 출력관리 하도록 리팩토링 했습니다.

[이 글](https://heesangstudynote.tistory.com/109)을 참조하여 고치려고 했는데 내용이 방대하고 잘 모르겠는 것도 많아서 어려움을 겪어 GPT와 함께 고쳐봤습니다..
Validate의 사용을 잘 이해 못해서 어려움을 겪고 있습니다 🥲

seulnan commented on src/main/java/mission/model/Fish.java:
자바에는 레코드라는 타입이 있는데 이거 알아보시면 좋을 것 같아요~

Shin-Yun-Cheol replied:
record타입은 [이 글](https://colevelup.tistory.com/28)과 GPT함께 보면서 공부해봤는데요 
불변 데이터 클래스를 간단하게 선언할 때 쓰는 문법입니다.
자동으로 equals, hashCode, toString, getter, 생성자를 만들어주는데 엄청난 문법이네요..!! 
Fish.java는 완벽하게 대체할 수 있을 것 같습니다.

seulnan commented on src/main/java/mission/model/FishWithCount.java:
```diff
@@ -0,0 +1,26 @@
+        return count;
+    }
+
+    public void eaten(int n){
```
오! 이렇게 fish의 상태와 eaten되는 동작을 함께 묶어서 모델링을 잘하셨어요! 
객체 자율성에 따른 협력 방식을 한번 공부해보시는 건 어떨까요?

읽어보기 좋은 글
https://gyuwon-tech.tistory.com/26

Shin-Yun-Cheol replied:
**인터페이스를 외부에 공개함으로써, 외부에서는 인터페이스를 통해 객체와 협력하고, 객체는 인터페이스 뒤에 숨어서 자율성을 보장받을 수 있다.**

이 말이 제일 와닿는 말인 것 같은데요. 이번 미션에서 인터페이스로 구현한게 없다보니 아직 확실히 알진 못하는 것 같아요. 
다만 이 코드에서처럼 FishWithCount에 eaten과 같은 동작을 같이 넣으면 다른 곳에서 이 동작을 구체적으로 어떻게가 아니라 무엇을 하라고만 정해주면, 그 동작을 하면 되기 때문에 각 객체들의 자율성이 오른다고 생각합니다!

seulnan replied:
음..저는 인터페이스로 구현과 로직을 분리하는것보다는 객체내에 책임을 어디까지 둬야하냐를 중점으로 이해하셨으면 좋겠는 마음에 저 글을 첨부했습니다.

제가 오브젝트 1장에서 인상깊게 읽은 문구인데요, 한번 읽어보시면 좋겠습니다.
아래 문구가 추상적이게 느껴진다면 오브젝트 1장에서는 실제 예시를 바탕으로 설명하니까 이를 읽어보시면 됩니다.

- 여러분의 코드에서 데이터와 데이터를 사용하는 프로세스가 별도의 객체에 위치하고 있다면 절차적 프로그래밍 방식을 따르고 있을 확률이 높다. 데이터와 데이터를 사용하는 프로세스가 동일한 객체 안에 위치한다면 객체지향 프로그래밍 방식을 따르고 있을 확률이 높다.
- ***객체가 어떤 데이터를 가지느냐보다는 객체에 어떤 책임을 할당할 것이냐에 초점을 맞춰야한다.***
- 불필요한 세부사항을 캡슐화하는 자율적인 객체들이 낮은 결합도와 높은 응집도를 가지고 협력하도록 최소한의 의존성만을 남기는 것이 훌륭한 객체지향 설계다.
- **_훌륭한 객체지향 설계란 소프트웨어를 구성하는 모든 객체들이 자율적으로 행동하는 설계를 가리킨다. 그 대상이 비록 실세계에서는 생명이 없는 수동적인 존재라고 하더라도 객체지향의 세계로 넘어오는 순간 그들은 생명과 지능을 가진 싱싱한 존재로 다시 태어난다._**

Shin-Yun-Cheol replied:
공부하다보니 구조를 무조건 객체와 유틸(?)으로 나눠야한다고 생각하고 있었는데, 혼자 책임질 수 있는 정도의 기능은 객체 안에 넣는 것이 객체지향적으로 더 자기 데이터를 보호하면서 자율적으로 행동할 수 있다고 배울 수 있었습니다!

seulnan commented on src/main/java/mission/util/FishInputParser.java:
이 파일에서 중첩된 예외처리가 많이 보이는데 이는 디버깅&테스트의 어려움을 유발해요.

위에서도 말했듯이 예외처리 관련한것은 validate폴더를 하나 따로 만들어서 거기에 몰아넣는것도 좋아보여요. 
validate 메서드를 따로 분리하고, 예외메시지는 enum 사용, 그리고 커스텀예외클래스를 만들면 더 명확한 예외처리가 가능해요.

Shin-Yun-Cheol replied:
[커밋링크](https://github.com/allrounder-backend/mission-food-chain/pull/3/commits/c3254f11c9340886f50282022b6795d2d655b8f2)

seulnan commented on src/main/java/mission/util/FishInputParser.java:
```diff
@@ -0,0 +1,50 @@
+            }
+
+            return map;
+        } catch (Exception e) {
```
여기서 try에서 구체적인 예외메시지들을 잘 던졌는데 이를 catch에서 잡아서 다시 "잘못된 입력"이라는 메시지로 뭉개버리면서 예외를 다시 던지네용..?

이러면 위에 적은 "입력에 유효한 항목이 없습니다." "숫자 형식이 잘못되었습니다: " 같은 구체적인 메시지 대신 "잘못된 입력"만 뜨게 되면서 구체적이지않은 예외처리가 되어버려요ㅠㅠ

Shin-Yun-Cheol replied:
에러처리 수정할 때 고쳐봤습니다. 
Exception 패키지에 ErrorCode와 InvalidException이라는 사용자 정의 예외를 이용했습니다!

seulnan replied:
아래 재리뷰확인부탁드립니다

seulnan commented on src/main/java/mission/Application.java:
```diff
@@ -1,9 +1,34 @@
+        );
+
+        FishInputView inputView = new FishInputView();
+        FishController controller = new FishController(data, inputView);
```
외부로부터 inputview가 주입받는게 아니라 직접 생성하는 것은 왜 좋지 않을까요??
DI와 객체생성 책임에 관해서 한번 공부해보시는 걸 추천드려요!

seulnan replied:
엥

Shin-Yun-Cheol replied:
**OCP**
OCP원칙은 확장에는 열려있어야하고 변경에는 닫혀있어야한다는 뜻입니다. 즉, 기존 코드를 건드리지 않고도, 새로운 기능을 확장할 수 있어야한다는 원칙입니다. 
자바에서 중요하다고 배운 추상화를 이용하는 개념이라고 이해했고, 인터페이스와 DI를 통해 지킬 수 있다고 공부했습니다.

**DI**
DI(Dependency Injection)은 객체를 직접 생성하지 않고 외부에서 주입 받는 것을 말합니다. 
이건 의존성을 외부에서 가지고 온다고 생각할 수 있습니다. 의존성은 클래스 A가 클래스 B의 객체를 참조할 때, A가 B에 의존성을 가진다고 합니다. 의존 관계를 인터페이스로 추상화한다면, 더 다양한 의존 관계(여러가지 클래스로 implement한다)를 맺을 수 있고 결합도가 낮아집니다.

이렇게 되면 생성자가 바뀌거나 인터페이스를 이용해서 더 다양한 객체를 만들 수 있고 테스트용 객체를 전달할 수도 있기 때문에 테스트 시에도 용이하다고 합니다.

찾아보니 위의 제 코드에서는 FishInputView와 FishController를 외부에서 연결해주는 Config패키지에 AppConfig같은 걸 만들어서 객체 생성 책임을 나눌 수 있을 것 같습니다.

다만 이렇게 되는 경우에 너무 많은 패키지와 클래스가 생기는게 아닌지, 아니면 원래 이정도로 나누는 것이 맞는지  궁금합니다.


[커밋 링크](https://github.com/allrounder-backend/mission-food-chain/pull/3/commits/c687044e2c1da62bc545e189862c1dd533fd5ab0)

seulnan replied:
넵 SOLID의 DIP도 추가적으로 공부해보시면 좋을 것 같아요 비슷한 얘기지만 좀더 구체적인 얘기입니다. 그저 구현이 아닌 인터페이스의 의존하거나 의존성을 단순히 주입하는 것뿐만아니라 잘못된 의존성주입은 뭔지, 의존성이 역전되는것에 대해서도 막아야하는걸 한번 고민해보셨으면 좋겠습니다.

seulnan reviewed (commented):
구현하시느라 고생많으셨습니다^0^

제 코멘트 아래에 
1. 자신이 공부한 것 (항상 공부할 땐 왜 써야하는지, 왜 이렇게 설계된건지, 어떻게 써야하는지를 중심으로 공부하시길 추천드립니다)
2. 리팩토링을 했다면 어떻게 했는지(관련 커밋코드 붙여주면 더 좋음!)

이정도로..? 전부 반영할 필요는 없지만 이정도는 공부하고 페플하면 더 뜻깊은 시간이 될 것 같습니다.

특히 DI 중심으로 한번 공부해보시면 MVC패턴이 이해되지않을까..말씀드려봅니다.(ex. 모델은 컨트롤러와 뷰에 의존하면안된다)

수고 많으셨습니다.

Shin-Yun-Cheol committed:
refactor : Application.java Factory 패턴 적용

Application.java에서 Simple Factory 패턴을 이용해서 객체 생성을 분리했습니다.

Shin-Yun-Cheol committed:
refactor : FishController 책임 분산

FishController.java에 있던 Fish객체와 FishwithCounts를 잇는 과정을 FishMapper라는 util클래스를 만들어서 의존성을 분리하고, Graph도 Factory에서 먹이사슬을 만들 수 있게 했습니다. 이 Factory를 통해 다른 먹이사슬 테스트도 할 수 있을 것으로 기대합니다.
OutputView.java를 추가하여 sout으로 출력하던 것을 output을 관리하는 view class를 하나 만들어서 처리했고 , 추후 이곳에서 에러처리도 할 수 있을 것입니다.

Shin-Yun-Cheol committed:
feat(Exception) : Exception패키지와 OuputView를 이용한 에러메시지 통합 관리

Exception패키지에 enum으로 에러메시지를 정의하고 InvalidException.java을 통해 예외 객체를 만들 수 있게 했습니다. FishController.java를 try-catch문으로 감싸서 에러 발생시 catch해서 OutputView에 정의해둔 메소드로 에러 출력하게 했습니다.

Shin-Yun-Cheol committed:
refactor(PredationGraph) : 먹이사슬 관계를 enum으로 정의

PredationGraphFactory.java에서 먹이사슬 관계를 하드코딩 하던 것을 enum을 이용해서 처리.

Shin-Yun-Cheol committed:
chore(rename) : service패키지 이름 소문자로 바꾸기

seulnan commented on src/main/java/mission/exception/ErrorCode.java:
```diff
@@ -0,0 +1,19 @@
+package mission.exception;
+
+public enum ErrorCode {
```
exception 이름을 정할땐 항상 에러범위가 되는 도메인(Fish, Input)을 앞에 붙여서 추상적이지않도록 하는게 좋은것같아요. 그게 확장성면에서도 좋구요.
예를들어, 현재 윤철코드처럼 단순히 포괄적인 ErrorCode라고해버리면 추후에 fishError를 추가하든, InputError를 추가하든, 삭제하든, 수정하든 할때 모두 이 파일에서 수정하게 되면 확장성이 떨어지겠죠??
실제로 enum을 쓸 때 주의해야할 점도 포괄적인 네이밍을 사용하면 다른 파일에서 enum을 사용한걸 보고 enum클래스를 다시 들어가서 확인해봐야하는 불편함이 생기기때문에 최대한 자세하게 적고 도메인별로 분리하는 것이 좋아보여요.
FishError나 InputError같은것으로 분리하게 된다면, 아 이사람은 fish에 대해, 글고 input에 대해 이런 에러메시지를 던지는구나가 바로 파악이 쉽게 되니까요.

seulnan commented on src/main/java/mission/exception/ErrorCode.java:
```diff
@@ -0,0 +1,19 @@
+
+    private final String message;
+
+    ErrorCode(String message) {
```
위에서도 얘기했듯이 이 ErrorCode네이밍이 포괄적인 것도 있지만, 실제로 프로젝트를 진행할 때는 에러코드와 메시지는 구분하는 편이에요. 
개발자들끼리의 커뮤니케이션에서 매번 얘기할때마다 "아 그 해당물고기존재하지않을때 발생하는 예외 있잖아요" 이런식으로 길게 얘기할 수 없으니까 약어처럼 Fish-001로 에러코드를 매겨서 "Fish-001이 발생했는데 확인부탁드립니다"이렇게 소통하는 경우가 많습니다. 그래서 에러를 처리할 땐 Code와 Message를 구분하는 것이 좋아요.

참고용 제가 플젝에서 실제로 쓰는 에러관련공통응답형식

{
	"timestamp": "2025-07-20T01:23:45",
	"status": 400,
	"errorCode": "Date-001",
	"message": "날짜 형식이 맞지 않습니다."
}

seulnan commented on src/main/java/mission/exception/InvalidException.java:
```diff
@@ -0,0 +1,13 @@
+package mission.exception;
+
+public class InvalidException extends RuntimeException{
```
에러처럼 이것도 단순히 포괄적으로 InvalidException보다 좀더 구체적인 이름으로 던지는게 코드가독성면에서도 아 걍 단순히 invalid한게 아니라 이런 상황에서 예외였구나가 바로 파악됩니다.

Shin-Yun-Cheol replied:
[커밋링크](https://github.com/allrounder-backend/mission-food-chain/pull/3/commits/c201b30a2c34b494cafd88844eb79078981a138e)

Shin-Yun-Cheol committed:
chore(rename) : ErrorCode.java InputErrorCode로 구체적인 이름으로 수정
